src/test/recovery/t/050_check_recovery_backup.pl

```diff
@@ -0,0 +1,158 @@
+	$standby->start(fail_ok => 1);
+
+	$standby->wait_for_log('database system is shut down');
+
+	# configure primary connection with replication slot
+	unlink "$standby_pgdata/recovery.signal";
+	unlink "$standby_pgdata/postgresql.auto.conf";
+	$standby->set_standby_mode;
+	open $conf, '>', "$standby_pgdata/postgresql.auto.conf";
+	print $conf "primary_conninfo = ";
+	print $conf "'user=postgres port=" . $primary_port;
+	print $conf " host=" . $primary_host;
+	print $conf " sslmode=prefer sslcompression=1 gssencmode=prefer";
+	print $conf " krbsrvname=postgres target_session_attrs=any";
+	print $conf " application_name=gp_walreceiver'\n";
+	print $conf "primary_slot_name = 'internal_wal_replication_slot'\n";
+	close $conf;
+
+	$standby->_update_pid(0);
+	$standby->start;
+}
+
+mkdir($archivedir);
+
+# Initialize primary
+$primary->init(allows_streaming => 1, extra => ['--data-checksums']);
+$primary->append_conf("postgresql.conf", "archive_mode = 'on'");
+$primary->append_conf("postgresql.conf", "archive_command = 'cp -rv %p $archivedir/'");
+$primary->start;
+$primary->safe_psql('postgres', 'CREATE DATABASE test_db');
+$primary->safe_psql('test_db', 'CREATE TABLE test AS SELECT generate_series(1,10)');
+$primary->safe_psql('postgres', 'CREATE ROLE postgres WITH LOGIN REPLICATION');
+
+create_backup_primary($basebackupdir1, 'backup_label1');
+$primary->stop('smart');
+restore_backup_primary($basebackupdir1, "recovery_target_name = 'backup_label1'\nrecovery_target_timeline = 'current'", '001 -> 012');
+wait_recovery_and_switch_wal_primary('001 -> 012');
+
+create_backup_primary($basebackupdir2, 'backup_label2');
+$primary->stop('smart');
+restore_backup_primary($basebackupdir2, "recovery_target_name = 'backup_label2'\nrecovery_target_timeline = 'current'", '012 -> 013');
+wait_recovery_and_switch_wal_primary('012 -> 013');
+
+$primary->stop('smart');
+restore_backup_primary($basebackupdir2, "recovery_target_name = 'backup_label2'\nrecovery_target_timeline = 'current'", '012 -> 023');
+wait_recovery_and_switch_wal_primary('012 -> 023');
+
+$primary->stop('smart');
+restore_backup_primary($basebackupdir1, "recovery_target_name = 'backup_label1'\nrecovery_target_timeline = 'current'", '101 -> 112');
+wait_recovery_and_switch_wal_primary('101 -> 112');
+
+$primary->stop('smart');
+restore_backup_primary($basebackupdir2, "recovery_target_name = 'backup_label2'\nrecovery_target_timeline = 'current'", '112 -> 113');
+wait_recovery_and_switch_wal_primary('112 -> 113');
+
+$primary->stop('smart');
+restore_backup_primary($basebackupdir1, "recovery_target_name = 'backup_label1'\nrecovery_target_timeline = 'current'", '101 -> 114');
+
+# wait for end of recovery and create replication slot
+$primary->poll_query_until('test_db', "SELECT true")
+  or die "Timed out while waiting for creating replication slot after recovery";
+$primary->safe_psql('test_db', "SELECT pg_create_physical_replication_slot('internal_wal_replication_slot')");
```

Comment:
What do these tests with `primary` check? They all pass successfully and without a patch! Only the test with `standby` below does not pass without a patch.

Reply:
These steps are needed to generate non-linear history of timelines.

Reply:
Could you please write a little more about this in the comments to the test?

Reply:
Added.